core/primitives/src/utils.rs

```diff
@@ -468,6 +469,17 @@ pub fn derive_eth_implicit_account_id(public_key: &Secp256K1PublicKey) -> Accoun
     format!("0x{}", hex::encode(&pk_hash[12..32])).parse().unwrap()
 }
 
+/// Returns a NEP-616 compliant deterministic account id.
+/// This is a NEAR-implicit account ID which is fully defined by its initial state.
+pub fn derive_near_deterministic_account_id(
+    state_init: &DeterministicAccountStateInit,
```

Comment:
Maybe accept ["lazily serialized"](https://github.com/mitinarseny/near-sdk-rs/blob/327ebe55e97e0f3f168679d0ed5b733a13658581/near-sdk/src/state_init.rs#L180-L186) version of `StateInit` similar to how it's done [here](https://github.com/mitinarseny/near-sdk-rs/blob/327ebe55e97e0f3f168679d0ed5b733a13658581/near-sdk/src/environment/env.rs#L1264)? This will expose API for smart-contracts to serialize [only once](https://github.com/mitinarseny/near-sdk-rs/blob/327ebe55e97e0f3f168679d0ed5b733a13658581/examples/sharded-fungible-token/wallet/src/lib.rs#L88-L102).

Reply:
Hm, I'm not sure if near-primitives is the right place to define optimizations like this. We should try to keep types in here as unopinionated as possible.

In the SDK, there can still be wrappers around it. Making these raw types easier to work with is exactly what the SDK does. I believe this should be done then and there, not here and now.

@mitinarseny Do you agree with my view?

Reply:
Oh, my bad, I thought `near-primitives` is a dependency of `near-sdk`, but it turns out to be only [used](https://github.com/near/near-sdk-rs/blob/7476cbe4328571322094f9e463f1dc9602933cb3/near-sdk/Cargo.toml#L40-L47) for `cfg(not(target_arch = "wasm32"))`.